internal/service/resource_pool_fetcher.go

```diff
@@ -0,0 +1,159 @@
+implied. See the License for the specific language governing permissions and limitations under the
+License.
+*/
+
+package service
+
+import (
+	"bytes"
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"log/slog"
+	"net/http"
+
+	"github.com/machinebox/graphql"
+	"github.com/openshift-kni/oran-o2ims/internal/data"
+	"github.com/openshift-kni/oran-o2ims/internal/k8s"
+	searchmodel "github.com/stolostron/search-v2-api/graph/model"
+)
+
+//go:generate mockgen -source=resource_pool_fetcher.go -package=service -destination=resource_pool_fetcher_mock.go
+type ResourcePoolFetcher interface {
+	FetchItems(ctx context.Context) (resources data.Stream, err error)
+}
+
+type resourcePoolFetcher struct {
+	logger       *slog.Logger
+	cloudID      string
+	backendURL   string
+	backendToken string
+	graphqlQuery string
+	graphqlVars  *searchmodel.SearchInput
+	resources    data.Stream
+}
+
+func NewResourcePoolFetcher(
+	logger *slog.Logger,
+	cloudID, backendURL, backendToken string,
+	graphqlQuery string, graphqlVars *searchmodel.SearchInput) ResourcePoolFetcher {
+	return &resourcePoolFetcher{
+		logger:       logger,
+		cloudID:      cloudID,
+		backendURL:   backendURL,
+		backendToken: backendToken,
+		graphqlQuery: graphqlQuery,
+		graphqlVars:  graphqlVars,
+	}
+}
+
+// FetchItems returns a data stream of O2 ResourcePools.
+// The items are converted from Clusters an Nodes fetched from the search API.
+func (r *resourcePoolFetcher) FetchItems(
+	ctx context.Context) (resourcePools data.Stream, err error) {
+	// Search Cluster and Nodes
+	resultArr, err := r.getSearchResults(ctx)
+	if err != nil {
+		return
+	}
+
+	// Extract 'related' items array
+	searchResult := resultArr[0].(map[string]any)
+	relatedArr, err := data.GetArray(searchResult, "related")
+	if err != nil {
+		return
+	}
+
+	// Convert response to json
+	items, err := json.Marshal(searchResult)
+	if err != nil {
+		return
+	}
+	itemsReader := bytes.NewReader(items)
+	related, err := json.Marshal(relatedArr[0])
+	if err != nil {
+		return
+	}
+	relatedReader := bytes.NewReader(related)
+
+	// Create reader for Clusters and Nodes
+	clusters, err := k8s.NewStream().
+		SetLogger(r.logger).
+		SetReader(itemsReader).
+		Build()
+	if err != nil {
+		return
+	}
+	nodes, err := k8s.NewStream().
+		SetLogger(r.logger).
+		SetReader(relatedReader).
+		Build()
+
+	// Transform Nodes to Resources
+	r.resources = data.Map(nodes, r.mapNodeItem)
+
+	// Transform Clusters to ResourcePools
+	resourcePools = data.Map(clusters, r.mapClusterItem)
+
+	return
+}
+
+func (r *resourcePoolFetcher) getSearchResults(ctx context.Context) (result []any, err error) {
+	tr := &http.Transport{
+		TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+	}
+	httpClient := &http.Client{Transport: tr}
+	client := graphql.NewClient(r.backendURL, graphql.WithHTTPClient(httpClient))
+
+	// Set GraphQL query and vars
+	request := graphql.NewRequest(r.graphqlQuery)
+	request.Var("input", *r.graphqlVars)
```

Comment:
What is the advantage of using the `graphql` package here? Can't we do this with a plain HTTP client? My concern is that it looks like this `graphql` package will store all the result set in memory before returning control. That means that we can't process them in an streaming approach. Is there any way to do that with the `graphql` package? Otherwise it negates the advantages of using streams.

Reply:
It mainly makes the code less clean without it. Can be removed.